Plugins: Adds support for URL params in plugin routes (#23503)
grafana/grafana: docs/sources/plugins/developing/auth-for-datasources.md

# Authentication for data source plugins

Grafana has a proxy feature that proxies all data requests through the Grafana backend. The main benefit of using the proxy is secure handling of credentials when authenticating against an external/third-party API. The Grafana proxy also adds [CORS](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS) headers to the proxied requests.

The proxy supports:

- [authentication with HTTP Headers]({{< relref "#api-key-http-header-authentication" >}}).
- [token authentication]({{< relref "#how-token-authentication-works" >}}) and can automatically renew a token for the user when the token expires.

## How the proxy works

> The user saves the API key/password on the plugin config page and it is encrypted (using the `secureJsonData` feature) and saved in the Grafana database. When a request from the data source is made, the Grafana Proxy will:

Reviewer: ```suggestion
The user saves the API key/password on the plugin config page and it is encrypted (using the `secureJsonData` feature) and saved in the Grafana database. When a request from the data source is made, the Grafana proxy will:
```

> 

Reviewer: Capitalize the first letter of each step.

1. intercept the original request sent from the data source plugin.
1. load the `secureJsonData` data from the database and decrypt the API key or password on the Grafana backend.
1. if using token authentication, carry out authentication and generate an OAuth token that will be added as an `Authorization` HTTP header to the requests (or alternatively it will add a HTTP header with the API key).
1. renew the token if it has expired.
1. after adding CORS headers and authorization headers, forward the request to the external API.

This means that users that access the data source config page cannot access the API key or password after they have saved it the first time and that no secret keys are sent in plain text through the browser where they can be spied on.

For backend authentication to work, the external/third-party API must either have an OAuth endpoint or that the API accepts an API key as a HTTP header for authentication.

> ## Encrypting Sensitive Data

Reviewer: ```suggestion
## Encrypt sensitive data
```

```

## Plugin routes

A plugin route describes where the intercepted request should be forwarded to and how to authenticate for the external API. You can define multiple routes that can match multiple external API endpoints.

You specify routes in the `plugin.json` file for your data source plugin. [Here is an example](https://github.com/grafana/azure-monitor-datasource/blob/d74c82145c0a4af07a7e96cc8dde231bfd449bd9/src/plugin.json#L30-L95) with lots of routes (though most plugins will just have one route).

When you build your URL to the third-party API in your data source class, the URL should start with the text specified in the path field for a route. The proxy will strip out the path text and replace it with the value in the URL field.

### Simple plugin route example

- If my code makes a call to URL `azuremonitor/foo/bar` with this code:

  ```js
  this.backendSrv.datasourceRequest({
    url: url,
    method: "GET",
  });
  ```

- and the plugin has this route:

  ```json
  "routes": [{
    "path": "azuremonitor",
    "method": "GET",
    "url": "https://management.azure.com"
  }]
  ```

> - then the Grafana proxy will transform the url from the original request into `https://management.azure.com/foo/bar`
> - finally, it will add CORS headers and forward the request to the new url. This example does not do any authentication.

Reviewer 2: ```suggestion
- then the Grafana proxy will transform the URL from the original request into `https://management.azure.com/foo/bar`
- finally, it will add CORS headers and forward the request to the new URL. This example does not do any authentication.
```

> The `method` parameter is optional. It can be set to a specific HTTP verb to provide more fine-grained control - for example you might have two plugin routes, one for GET requests and one for POST requests.

Reviewer: ```suggestion
The `method` parameter is optional. It can be set to a specific HTTP verb to provide more fine-grained control. For example you might have two plugin routes, one for GET requests and one for POST requests.
```

### Dynamic routes

When using routes, you can also reference a variable stored in JsonData or SecureJsonData which is interpolated (replacing the variable text with a value) when the data source makes a request to the proxy. These are variables that were entered by the user on the data source configuration page and saved in the Grafana database.

In this example, the value for `dynamicUrl` comes from the JsonData blob and the api key's value is set from the SecureJsonData blob. The `urlParams` field is for query string parameters for HTTP GET requests.

```json
"routes": [
  {
      "path": "custom/api/v5/*",
      "method": "GET",
      "url": "{{.JsonData.dynamicUrl}}",
      "urlParams": [
        {"name": "apiKey", "content": "{{.SecureJsonData.apiKey}}"}
      ]
  }
]
```

Given that:

- `JsonData.dynamicUrl` has the value `http://example.com/api`
- `SecureJsonData.apiKey` has the value `secretKey`

> a call to the url: `custom/api/v5/some/path`

Reviewer 2: ```suggestion
a call to the URL: `custom/api/v5/some/path`
```

> will be proxied to the following url: `http://example.com/api/some/path?apiKey=secretKey`

Reviewer 2: ```suggestion
will be proxied to the following URL: `http://example.com/api/some/path?apiKey=secretKey`
```

}
```

> This interpolates in data from both `jsonData` and `secureJsonData` to generate the token request to the third-party API. It is common for tokens to have a short expiry period (30 minutes). The Grafana proxy will automatically renew the token if it has expired.

Reviewer: ```suggestion
This interpolates in data from both `jsonData` and `secureJsonData` to generate the token request to the third-party API. It is common for tokens to have a short expiry period (30 minutes). The Grafana proxy automatically renews the token if it has expired.
```
